src/util/BlankNodeManager.cpp

```diff
@@ -59,12 +77,142 @@ bool BlankNodeManager::LocalBlankNodeManager::containsBlankNodeIndex(
+  AD_CONTRACT_CHECK(blocks_->blocks_.empty() && otherBlocks_.empty(),
+                    "Explicit reserving of blank node blocks is only allowed "
+                    "for empty `LocalBlankNodeManager`s");
+
+  // Lambda that does the registering for a single `Blocks` object.
+  auto allocateSingleSet = [this](const OwnedBlocksEntry& entry) {
+    auto [blocks, isNew] =
+        blankNodeManager_->registerBlocksWithExplicitUuid(entry.uuid_);
+    // If the block is not new, then the `Blocks` with the given UUID have
+    // already been reserved by another `LocalBlankNodeManager` that owns the
+    // same set. If it is new, we are the first, and have to also allocate the
+    // blocks.
+    // TODO<joka921> It is ugly to not have this under a single lock and sublte,
+    // maybe this complete lambda should be a member function of the
+    // `BlankNodeManager` under a single lock.
+    if (isNew) {
+      for (const auto& idx : entry.blockIndices_) {
+        blocks->blocks_.emplace_back(
+            blankNodeManager_->allocateExplicitBlock(idx));
+      }
+    }
+    return blocks;
+  };
+
+  // The semantics of the argument is (as enforced by the `getOwnedBlockIndices`
+  // function): The first element is the `blocks_` primarily owned by this
+  // `LocalBlankNodeManager`, The remaining elements are the `otherBlocks_`.
+  AD_CONTRACT_CHECK(!indices.empty());
+  blocks_ = allocateSingleSet(indices.at(0));
+  otherBlocks_.reserve(indices.size() - 1);
+  for (const auto& entry : indices | ql::views::drop(1)) {
+    otherBlocks_.push_back(allocateSingleSet(entry));
+  }
+
+  // The following code ensures that the next call to `getId` allocates a new
+  // block. This is necessary, because we don't have access to the information
+  // which IDs in the allocated blocks actually are in use.
+  if (!blocks_->blocks_.empty()) {
+    idxAfterCurrentBlock_ = blocks_->blocks_.back().nextIdx_;
+  }
+}
+
+// _____________________________________________________________________________
+auto BlankNodeManager::createBlockSet() -> std::shared_ptr<Blocks> {
+  auto res = std::make_shared<Blocks>(this);
+  // Guard against the (very very unlikely) case of UUID collision
+  auto lock = state_.wlock();
+  while (true) {
+    auto [it, isNew] = lock->managedBlockSets_.try_emplace(res->uuid_, res);
+    if (isNew) {
+      return res;
+    }
+    // Note: the only realistic way to cover this is by tampering with the UUID
+    // generator.
+    res->uuid_ = lock->uuidGenerator_();
+  }
+}
+
+// _____________________________________________________________________________
+void BlankNodeManager::freeBlockSet(Blocks& blocks) {
+  // We keep the lock the whole time to not make any inconsistent state visible
+  // to the outside world.
+  auto lock = state_.wlock();
+  // First unregister the UUID.
+  auto it = lock->managedBlockSets_.find(blocks.uuid_);
+  AD_CORRECTNESS_CHECK(it != lock->managedBlockSets_.end());
```

Comment:
I think it's better to return instead of throwing an exception if this happens.
Consider the case where
1. `registerBlocksWithExplicitUuid` is called
2. The reference counter of the shared pointer reaches 0 and thus `freeBlockSet`, but doesn't get the lock for now.
3. `registerBlocksWithExplicitUuid` is called again using the same UUID, get's the lock for some reason while the other thread is still waiting.
4. The reference count again goes to 0 and calls `freeBlockSet`, but this thread gets the lock and removes itself succesfully
5. Now the first thread gets the lock and suddenly there's no longer an entry in the list and we get the exception.

Incredibly unlikely but possible.